lib/web.ts

```diff
@@ -249,11 +249,17 @@ export default class AutoConsent {
         this.updateState({ findCmpAttempts: this.state.findCmpAttempts + 1 });
         const foundCMPs: AutoCMP[] = [];
 
-        for (const cmp of this.rules) {
+        const isTop = window.top === window;
+        // refilter relevant rules for this context
+        const rulesForContext = this.rules.filter((cmp) => {
+            if (isTop) {
+                return cmp.runContext.main || cmp.runContext === undefined;
+            }
+            return cmp.runContext.frame || false;
+        });
```

Comment:
Looking at the current implementation, we're iterating over all rules and checking runContext multiple times (lines 254, 279, 287)
I feel like you can optimize more by grouping rulesets in this first pass. What do you think?

Reply:
I took another pass at this, let me know what you think.